support-workers/src/main/scala/com/gu/emailservices/DigitalPackEmailFields.scala

```diff
@@ -1,78 +1,254 @@
+    }
+
+    case class DDAttributes(
+      account_number: String,
+      sort_code: String,
+      account_name: String,
+      mandateid: String,
+      default_payment_method: String = "Direct Debit",
+    ) extends PaymentFieldsAttributes
+
+    case class CCAttributes(
+      default_payment_method: String = "Credit/Debit Card",
+    ) extends PaymentFieldsAttributes
+
+    case class PPAttributes(
+      default_payment_method: String = "PayPal",
+    ) extends PaymentFieldsAttributes
+
+  }
+
+
+  case class BasicDSAttributes(
+    zuorasubscriberid: String,
+    emailaddress: String,
+    first_name: String,
+    last_name: String,
+    subscription_details: String,
+  ) extends DigitalSubscriptionEmailAttributes
+
+  case class DirectDSAttributes(
+    directCorp: BasicDSAttributes,
+    subscription_term: String,
+    payment_amount: String,
+    country: String,
+    date_of_first_payment: String,
+    currency: String,
+    trial_period: String,
+    paymentFieldsAttributes: PaymentFieldsAttributes,
+  ) extends DigitalSubscriptionEmailAttributes
+
+  case class GifteeRedemptionAttributes(
+    gift_recipient_first_name: String,
+    subscription_details: String,
+    gift_start_date: String,
+    gift_recipient_email: String,
+  ) extends DigitalSubscriptionEmailAttributes
+
+  case class GifterPurchaseAttributes(
+    gifter_first_name: String,
+    gifter_last_name: String,
+    gift_recipient_first_name: String,
+    gift_recipient_last_name: String,
+    gift_recipient_email: String,
+    gift_personal_message: String,
+    gift_code: String,
+    gift_delivery_date: String,
+    subscription_details: String,
+    paymentAttributes: PaymentFieldsAttributes,
+    date_of_first_payment: String,
+  ) extends DigitalSubscriptionEmailAttributes
+
+  case class GifteeNotificationAttributes(
+    gifter_first_name: String,
+    gift_personal_message: String,
+    gift_code: String,
+  ) extends DigitalSubscriptionEmailAttributes
 
-    val fields = List(
-      "zuorasubscriberid" -> subscriptionNumber,
-      "emailaddress" -> user.primaryEmailAddress,
-      "first_name" -> user.firstName,
-      "last_name" -> user.lastName,
-    ) ++ fieldsForReaderType
+}
+class DigitalPackEmailFields(
+  subscriptionEmailFields: SubscriptionEmailFields,
+) {
+
+  import DigitalPackEmailFields._
+  import DigitalSubscriptionEmailAttributes._
+  import subscriptionEmailFields._
+  import allProductsEmailFields._
+
+  private def directOrCorpFields(details: String) = BasicDSAttributes(
+    zuorasubscriberid = subscriptionNumber,
+    emailaddress = user.primaryEmailAddress,
+    first_name = user.firstName,
+    last_name = user.lastName,
+    subscription_details = details
+  )
 
-    val dataExtensionName = if (paidSubPaymentData.isDefined) "digipack" else "digipack-corp"
+  def build(
+    paidSubPaymentData: Option[PaymentMethodWithSchedule],
+    readerType: ReaderType
+  ): Either[String, List[EmailFields]] = {
+
+    val Purchase = Some
```

Comment:
This is really nice 👏

Reply:
while I was looking it up, I found scala 3 has a really middle ground between type alias and value class, which should reduce the boilerplate for impementing a new type just for a bit of validation.
http://dotty.epfl.ch/docs/reference/other-new-features/opaques.html